### ISSUE TITLE
Document-dependent path evaluation

### ISSUE THREAD

**Issue by jakajancar** [halaxa/json-machine#102]
Let's say we have a property with an array of dynamically-typed, user-provided parameters.

```php
$options = ['pointer' => '/user-provided-parameters/-'];

Items::fromString('{"user-provided-parameters": [1,2,3]}', $options);
// Expected: [1,2,3]
// Actual: same

Items::fromString('{"user-provided-parameters": [1,false,null]}', $options);
// Expected: [1,false,null]
// Actual: same

Items::fromString('{"user-provided-parameters": [1, [2,3]]}', $options);
// Expected: [1, [2,3]]
// Actual: [1, 2, 3]

Items::fromString('{"user-provided-parameters": []}', $options);
// Expected: []
// Actual: Exception: Paths '/user-provided-parameters/-' were not found in json stream.
```

This makes JSON Machine not very useful for working with documents with a more dynamic schema. Moreover, even arrays have a special case at length == 0.

I checked the JSON Pointer spec to see if this is an implementation bug or by design. Seems like JSON Pointer is not intended for the (JSONPath-like) selection at all, but for navigation to a single node. Even the `-` is interpreted differently (`the (nonexistent) member after the last array element` vs `a wildcard which matches any array index`). It also would not have the above problem and would always navigate to the expected subtree. It would be better if the readme said "a syntax inspired by JSON Pointer".

Re. a solution, it would be great if there was an option to not automatically descend deeper than the specified path and make the subtree selection not dependent on the values in it.

**Comment by halaxa** [comment]
TL;DR Just omit the dash at the end.

Hi. Your example works as expected. It seems in your case the JSON Pointer (`pointer` option) is just not used correctly. The `pointer` option means "iterate over items in this element". If you only need to iterate over the items in the `user-provided-parameters` key, just use `/user-provided-parameters` as the pointer. The dash at the end means "any index" so it matches `/user-provided-parameters/0`, `/user-provided-parameters/1`, and so on, and then tries to iterate over what's inside a vector on that index. If you need more explanation, let me know or have a second look at the JSON Machine documentation.

**Comment by halaxa** [comment]
I'm not sure what the question is now. Can you be more specific?

Anyway, let me just elaborate a little on the `flatten` topic. JSON Machine supports finding data in a JSON down to a single scalar value if needed. It does that automatically. If it finds a scalar value at a pointer instead of an object or an array, it just yields it in a single iteration. So it might seem it somehow flattens the structure when used in combination with `-` and when the structure is not rigid. But in reality, no such thing happens.

Try this and you'll see no deep flattening is happening:
```php
$options = ['pointer' => '/table/-'];

Items::fromString('{"table": [[[1,2]], [3,4]]}', $options);
// Expected: [[1,2],3,4]
```

Also, this example is not expected to produce an error:
```php
$options = ['pointer' => '/table/-'];
Items::fromString('{"table": [[1,2], 3]}', $options);
```
because at `/table/0` there is `[1,2]` which is sequentially iterated, and at `/table/1` there is `3` which is a scalar value and as such it's simply yielded as a single value.

**Comment by halaxa** [comment]
> - If you use `/2d/-/-`
>   - ✅ You do get both indices.
>   - ❌ Third valid example ([[1,2], [3,[4,5]]]) gets flattened (and you get 5 items)
>      - **That's a feature, not a bug as explained earlier.**
>   - ✅ The invalid example ignores the invalid element.
>
> - If you use `/2d/-`:
>   - ❌ You do not get both indices, only the first.
>      - **Ok, this seems weird. Can you give an example? Can it be the same problem as #100?**
>   - ✅ Third valid example doesn't get flattened (properly get 4 elements)
>   - ❌ The invalid example gets silently ignored (you get same items as first valid example)
>      - **Not-found items get ignored. That's normal behavior. It's as if you wanted the `find` command to fail on every existing file in the searched dir that does not match searched string.**

**Comment by halaxa** [comment]
Sorry for being brief ;)

**Comment by halaxa** [comment]
> But disabling this feature is essentially my feature request! :D

Now it makes perfect sense 😁. Because in terms of JSON Machine, there's no 'flattening', I'd suggest modifying the scalar parsing logic, which is what's actually behind your problem. Maybe an option something like `iterate_scalars`, with three settings:
- `AUTO` (current behavior, would remain the default)
- `ALWAYS`/`ONLY`/`FORCE` (an iterable on the pointer position will throw)
- `NEVER` (a scalar on the pointer position will throw)

This example of yours:
> $options = ['pointer' => '/table/-'];
> 
> Items::fromString('{"table": [[1,2], 3]}', $options);
> // Expected: error
> // Actual: [1,2,3]

would then throw an error with option `'iterate_scalars' => NEVER`

**Comment by halaxa** [comment]
Also for a less predictable structure maybe #36 would help?

**Comment by halaxa** [comment]
What do you think about the solution proposed above? (`iterate_scalars` option as a feature request)?

**Comment by halaxa** [comment]
As for the recursive iteration (#36), it is now finished, merged, and released.
